README.md
# appa-api

Quickly create simple HTTP services.

[![npm][npm-image]][npm-url]
[![travis][travis-image]][travis-url]
[![standard][standard-image]][standard-url]
[![conduct][conduct]][conduct-url]

[npm-image]: https://img.shields.io/npm/v/appa-api.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/appa-api 
[travis-image]: https://img.shields.io/travis/sethvincent/appa-api.svg?style=flat-square
[travis-url]: https://travis-ci.org/sethvincent/appa-api
[standard-image]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square
[standard-url]: http://npm.im/standard
[conduct]: https://img.shields.io/badge/code%20of%20conduct-contributor%20covenant-green.svg?style=flat-square
[conduct-url]: CONDUCT.md

![appa](https://raw.githubusercontent.com/sethvincent/appa-api/master/appa.jpg)

## Install

Make sure you've got [node installed](http://nodejs.org), then make `appa` a project dependency:

```sh
npm install --save appa
```

## Usage

```js
var http = require('http')
var app = require('appa-api')()
var send = require('appa-api/send')
var log = app.log

app.on('/', function (req, res, context) {
  send(res, { message: 'oh hey friends' })
})

http.createServer(app).listen(3000, function () {
  log.info('server started at http://127.0.0.1:3000')
})
```

## Error handling

Any uncaught errors that occur in a request handler will be caught and a `500 Internal server error` response will be sent.

Send error responses using the `appa-api/error` module:

```js
var error = require('appa-api/error')

module.exports = function (req, res, ctx) {
  return error(res, 404, 'Not found')
}
```

Sending an error response does not automatically log the error, so to add that you can do something like:

```js
var error = require('appa-api/error')
var log = require('appa-api/log')()

module.exports = function (req, res, ctx) {
  log.error(req.method, '500', errorStack)
  return error(res, 500, 'Internal server error')
}
```

## Logging

appa uses [pino](https://npmjs.com/pino) for logging. Pass options to pino with `options.log`: `appa({ log: pinoOptions })`.

See [example pino usage](https://github.com/pinojs/pino#usage) and [all pino options](https://github.com/pinojs/pino/blob/master/docs/API.md#pinooptions-stream).

Or disable logging completely by setting `options.log` to `false`: `appa({ log: false })`.

## Documentation
- [API](docs/api.md)
- [Tests](tests/)

### Examples
- [Basic example](examples/basic.js)
- Example [server](examples/server.js) and [client](examples/client.js)

## Contributing

Contributions are welcome! Please read the [contributing guidelines](CONTRIBUTING.md) first.

## Conduct

It is important that this project contributes to a friendly, safe, and welcoming environment for all. Read this project's [code of conduct](CONDUCT.md)

## Changelog

Read about the changes to this project in [CHANGELOG.md](CHANGELOG.md). The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## Contact

- **issues** – Please open issues in the [issues queue](https://github.com/sethvincent/appa-api/issues)
- **twitter** – Have a question? [@sethdvincent](https://twitter.com/sethdvincent)

## License

[ISC](LICENSE.md)


CHANGELOG.md
# appa-api change log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## Unreleased

> Note: unreleased changes are listed here

## 7.0.0 - 2017-04-30

## Changed
- renamed this project from appa to appa-api
- now uses the [send-response](https://github.com/sethvincent/send-response) module
- as a result, switched back to `send(res, data)` signature instead of `send(data).pipe(res)`

## 6.1.2 - 2017-02-25

## Fixed

- fix double response bug

## 6.1.1 - 2017-02-25

## Fixed

- uncaught errors inside request handlers are now caught and a `500 Internal server error` response is sent
- Added docs to readme about error handling and logging
- update tests/README.md

## 6.1.0 - 2017-02-19

## Added
- disable logging with `var app = appa({ logging: false })`
- Improved [api docs](docs/api.md)
- refactored to use the raw-body module, which allows setting the limit of incoming request body sizes
- individual request handlers can now pass a `parseJSON: false` option to disable JSON parsing
- add [server](examples/server.js) and [client](examples/client.js) examples

## 6.0.0 - 2016-11-01

### Changed
- use `send(data).pipe(res)` rather than `send(res, data)`

## 5.0.1 - 2016-10-24

### Added
- Add a docs directory
- Add CONDUCT.md
- Add CHANGELOG.md

### Changed
- Move send, error, and log into separate files
- Revise README.md with links and updated info.
- License: MIT > ISC
- Create tests directory


tests/index.js
var test = require('tape')
var http = require('http')
var request = require('request')
var intoStream = require('into-stream')

var createApp = require('../index')
var send = require('../send')

function createServer (app) {
  return http.createServer(app)
}

test('create a server', function (t) {
  var app = createApp({ log: false })
  var server = createServer(app).listen(0, function () {
    t.ok(app)
    server.close()
    t.end()
  })
})

test('create a route', function (t) {
  t.plan(6)
  var app = createApp({ log: false })

  app.on('/', function (req, res, context) {
    t.ok(req)
    t.ok(res)
    t.ok(context)
    send(res, { hello: 'hi' })
  })

  var server = createServer(app).listen(3131, function () {
    request({ url: 'http://127.0.0.1:3131', json: true }, function (err, res, body) {
      t.notOk(err)
      t.ok(body)
      t.equal(body.hello, 'hi')
      server.close()
    })
  })
})

test('querystring is parsed', function (t) {
  t.plan(4)
  var app = createApp({ log: { level: 'silent' } })

  app.on('/', function (req, res, context) {
    send(res, context.query)
  })

  var server = createServer(app).listen(3131, function () {
    request({ url: 'http://127.0.0.1:3131?hi=hello&hey[wut]=wat', json: true }, function (err, res, body) {
      t.notOk(err)
      t.ok(body)
      t.equal(body.hi, 'hello')
      t.equal(body.hey.wut, 'wat')
      server.close()
    })
  })
})

test('send a stream', function (t) {
  t.plan(4)
  var app = createApp({ log: false })

  app.on('/', function (req, res, context) {
    var stream = intoStream.obj(context.query)
    send(res, stream)
  })

  var server = createServer(app).listen(3131, function () {
    request({ url: 'http://127.0.0.1:3131?hi=hello&hey[wut]=wat', json: true }, function (err, res, body) {
      t.notOk(err)
      t.ok(body)
      t.equal(body.hi, 'hello')
      t.equal(body.hey.wut, 'wat')
      server.close()
    })
  })
})

test('receive params', function (t) {
  t.plan(7)
  var app = createApp({ log: false })

  app.on('/list/:listkey/item/:itemkey', function (req, res, ctx) {
    t.ok(ctx.params)
    t.ok(ctx.params.listkey)
    t.ok(ctx.params.itemkey)
    send(res, ctx.params)
  })

  var server = createServer(app).listen(3131, function () {
    request({ url: 'http://127.0.0.1:3131/list/hello/item/wat', json: true }, function (err, res, body) {
      t.notOk(err)
      t.ok(body)
      t.equal(body.listkey, 'hello')
      t.equal(body.itemkey, 'wat')
      server.close()
    })
  })
})

test('handle internal errors', function (t) {
  t.plan(4)
  var app = createApp({ log: false })

  app.on('/error', function (req, res, ctx) {
    throw new Error('oops')
  })

  var server = createServer(app).listen(3131, function () {
    request({ url: 'http://127.0.0.1:3131/error', json: true }, function (err, res, body) {
      t.notOk(err)
      t.equal(res.statusCode, 500)
      t.equal(body.statusCode, 500)
      t.equal(body.message, 'Internal server error')
      server.close()
    })
  })
})


index.js
var assert = require('assert')
var qs = require('qs')
var url = require('url')
var rawBody = require('raw-body')
var createRouter = require('wayfarer')
var isType = require('type-is')
var httplogger = require('pino-http')
var jsonParse = require('fast-json-parse')

var createLogger = require('./log')
var send = require('./send')
var error = require('./error')

/**
* Create the application. Returns the `app` function that can be passed into `http.createServer`.
* @name createApp
* @param {Object} config
* @param {Object|Boolean} config.log – Set to `false` to disable logging. Set to an object with options that are passed to [pino](https://npmjs.com/pino) and [pino-http](https://npmjs.com/pino-http)
* @param {Function} config.notFound – Function that handlers 404 routes. Provides `req`, `res`, and `ctx` arguments just like other appa route handlers. Default: a function that sends a `404` statusCode with the JSON `{ message: 'Not found' }`.
* @returns {Function} `app`
* @example
* var appa = require('appa-api')
*
* var app = appa({
*   log: { level: 'info' }, // or set to `false` to disable all logging
*   notFound: function (req, res, ctx) {
*     return app.error(res, 404, 'Not Found')
*   }
* })
*/
module.exports = function createApp (config) {
  config = config || {}

  config.log = config.log === false
    ? { level: 'silent' }
    : config.log || { level: 'info' }

  config.log.name = config.log.name || 'appa'

  var log = createLogger(config.log)
  var httplog = httplogger(config.log, config.log.stream)
  var router = app.router = createRouter('/404')

  // provide a 404 fallback
  on('/404', (config.notFound || notFound))
  function notFound (req, res) { error(res, 'Not found') }

  // ignore favicon.ico requests
  on('/favicon.ico', function (req, res) { send(res, 200) })

  /**
  * The request, response handler that is passed to `http.createServer`, and the object that
  * provides methods for your app.
  * @name app
  * @param {Object} req – the http request object
  * @param {Object} res – the http response object
  * @example
  * var http = require('http')
  * var appa = require('appa-api')
  *
  * var app = appa()
  * var server = http.createServer(app)
  */
  function app (req, res) {
    httplog(req, res)
    var ctx = url.parse(req.url)
    ctx.query = qs.parse(ctx.query)
    return router(ctx.pathname, req, res, ctx)
  }

  /**
  * Route handler
  * @name app.on
  * @param {String} pathname – the route for this handler
  * @param {Object} options – options for the request handler
  * @param {Boolean} options.parseJSON – optionally disable JSON parsing of the body. Default: `true`
  * @param {Function} callback – the route handler
  * @example
  * app.on('/', function (req, res, ctx) {
  *   // handle the route
  * })
  */
  function on (pathname, options, callback) {
    if (typeof options === 'function') {
      callback = options
      options = {}
    }

    assert.equal(typeof pathname, 'string', 'appa: pathname is required and must be a string')
    assert.equal(typeof options, 'object', 'appa: options must be an object')
    assert.equal(typeof callback, 'function', 'appa: callback function is required')

    options.parseJSON = options.parseJSON === false ? options.parseJSON : true

    return router.on(pathname, function (params, req, res, ctx) {
      ctx.params = params

      function respond (req, res, ctx) {
        try {
          return callback(req, res, ctx)
        } catch (err) {
          log.error(err)
          return error(res, 500, 'Internal server error', err)
        }
      }

      if (req.method === 'POST' || req.method === 'PUT' || req.method === 'PATCH') {
        body(req, options, function (err, result) {
          if (err) return error(res, err.status, err.type)

          if (isType(req, ['json']) && options.parseJSON) {
            ctx.body = parseJSON(res, result)
          } else {
            ctx.body = result
          }

          return respond(req, res, ctx)
        })
      } else {
        return respond(req, res, ctx)
      }
    })
  }

  function parseJSON (res, result) {
    try {
      return jsonParse(result)
    } catch (err) {
      log.error(err)
      return error(res, 400, 'Invalid JSON')
    }
  }

  function body (req, options, callback) {
    if (!options.limit) options.limit = '1mb'
    if (!options.encoding) options.encoding = 'utf8'
    rawBody(req, options, callback)
  }

  /**
  * Send a JSON object as a response
  * @name app.send
  * @param {Number} statusCode – the status code of the response, default is 200
  * @param {Object} data – the data that will be stringified into JSON
  * @example
  * var send = require('appa-api/send')
  *
  * app.on('/', function (req, res, ctx) {
  *   send(res, { message: 'hi' })
  * })
  */
  app.send = send

  /**
  * Send a JSON error response
  * @name app.error
  * @param {Number} statusCode – the status code of the response, default is 404
  * @param {String} message – the message that will be stringified into JSON
  * @param {Object} data – additional data about the error to send in the response
  * @example
  * var error = require('appa-api/error')
  *
  * app.on('/', function (req, res, ctx) {
  *   error(res, 404, 'Resource not found')
  * })
  */
  app.error = error

  /**
  * Create logs using the pino module: https://npmjs.com/pino
  * @name app.log
  */
  app.log = log

  /**
  * Parse or stringify a JSON stream using the JSONStream module: https://npmjs.com/JSONStream
  * @name app.json
  */
  app.json = require('JSONStream')

  /**
  * Compose a stream using the pump module: https://npmjs.com/pump
  * @name app.pipe
  */
  app.pipe = require('pump')

  app.on = on
  return app
}
